### ISSUE TITLE
Write Aggregation Queries in PHP

### ISSUE THREAD

**Issue by JugalSA** [elastic/elasticsearch-php#196]
Hello, i am writing Elasticsearch aggregations queries to find the total count available:

GET  listjugal/_search
{
    "aggregations": {
      "Brand_Name_Count": {
         "terms": {"field": "brand_name", "size" : 0}
      },
     "Stock_Status_Count" : {
           "terms" : { "field" : "stock_status", "size" : 50}
      },
     "Category_Id_Count" : {
           "terms" : { "field" : "category_id", "size" : 50}
      },
     "Price_Name_Count" : {
           "terms" : { "field" : "price_name", "size" : 50}
      },
     "Discount_Name_Count" : {
           "terms" : { "field" : "discount_name", "size" : 50}
      }
   }
}

And I am getting the count properly.
How do i write these type of queries in php code??
And also how do i merge aggregation with search results query.
Help me

**Comment by polyfractal** [comment]
The agg (and search) PHP syntax follows the JSON API 1:1.  So you can take your aggregation above and just translate it into PHP arrays like so:

```php
$myQuery = [];  // Your query goes here

$params = [
  'index' => 'listjugal',
  'body' => [
    'aggs' => [
      'Brand_Name_Count' => [
        'terms' => [
          'field' => 'brand_name',
          'size' => 0
        ]
      ],
      'Stock_Status_Count' => [
        'terms' => [
          'field' => 'stock_status',
          'size' => 50
        ]
      ],
      'Category_Id_Count' => [
        'terms' => [
          'field' => 'category_id',
          'size' => 50
        ]
      ],
      'Price_Name_Count' => [
        'terms' => [
          'field' => 'price_name',
          'size' => 50
        ]
      ],
      'Discount_Name_Count' => [
        'terms' => [
          'field' => 'discount_name',
          'size' => 50
        ]
      ]
    ],
    'query' => $myQuery
  ]
];

$results = $client - > search($params);
```

Aggregations are executed in parallel to searches, so just specify your search query and you'll get back a search `hits` element as well as the `aggs` element

**Comment by JugalSA** [comment]
Hi..Thanks for your help but i have one more issue related to Elaasticsearch json arrays and php arrays..
I want to form this array in php..
GET  purplle/_search
{
    "query" : {
      "filtered" : { 
         "filter" : {
            "bool" : {
              "should" : [
            {"term": {
                "name": "axe"
             }},
             {"term": {
                "meta_keywords": "axe"
             }},
             {"term": {
                "brand_name": "axe"
             }},
             {"term": {
                "category_name": "axe"
             }}
              ]
           }
         }
      }
   }
}


I have also tried this in php..below code
$filters = array();
$filters['bool']['should']['term']['name'] = 'axe';
$filters['bool']['should']['term']['meta_keywords'] = 'axe';
$filters['bool']['should']['term']['brand_name'] = 'axe';
$filters['bool']['should']['term']['category_name'] = 'axe';

$myQuery = array();
$myQuery['filtered'] = array(
   "filter" => $filters
	);
$params['body'] = array(
				'query' => $myQuery
			);

Which is giving me this output:

{
   "index":"purplle",
   "type":"cosmetics",
   "body":{
      "aggregations":{ },
      "query":{
         "filtered":{
            "filter":{
               "bool":{
                  "should":{
                     "term":{
                        "name":"axe",
                        "meta_keywords":"axe",
                        "brand_name":"axe",
                        "category_name":"axe"
                     }
                  }
               }
            }
         }
      }
   }
}

But I dont want this.There is only one term filter, i want four term filter in php like ElasticSearch Array.
It is not giving me any error, but results are not what i am expecting.
Please help soon if possible..

**Comment by JugalSA** [comment]
Hi..Thanks for your help but i have one more issue related to Elaasticsearch json arrays and php arrays..
I want to form this array in php..
GET purplle/_search
{
"query" : {
         "filtered" : { 
            "filter" : {
               "bool" : {
                 "should" : [
                      {"term": {
                          "name": "axe"
                      }},
                     {"term": {
                         "meta_keywords": "axe"
                      }},
                      {"term": {
                        "brand_name": "axe"
                         }},
                      {"term": {
                         "category_name": "axe"
                        }}
                     ]
                   }
                 }
               }
             }
              }

I have also tried this in php..below code
$filters = array();
$filters['bool']['should']['term']['name'] = 'axe';
$filters['bool']['should']['term']['meta_keywords'] = 'axe';
$filters['bool']['should']['term']['brand_name'] = 'axe';
$filters['bool']['should']['term']['category_name'] = 'axe';

$myQuery = array();
$myQuery['filtered'] = array(
"filter" => $filters
);
$params['body'] = array(
'query' => $myQuery
);

Which is giving me this output:

{
"index":"purplle",
"type":"cosmetics",
"body":{
       "query":{
        "filtered":{
          "filter":{
            "bool":{
              "should":{
                 "term":{
                    "name":"axe",
                   "meta_keywords":"axe",
                   "brand_name":"axe",
                   "category_name":"axe"
                }
             }
          }
        }
     }
  }
 }
}

But I dont want this.There is only one term filter, i want four term filter in php like ElasticSearch Array.
It is not giving me any error, but results are not what i am expecting.
Please help soon if possible..

**Comment by polyfractal** [comment]
You are overwriting the same key in the map each time you add a new term.  E.g. the key `['bool']['should']['term']` is repeatedly changed.

What you want instead is to append new array objects to that key, rather than overwriting it.  Add a `[]` in between the `should` and the `term` and it will append new objects:

```php
filters = array();
$filters['bool']['should'][]['term']['name'] = 'axe';
$filters['bool']['should'][]['term']['meta_keywords'] = 'axe';
$filters['bool']['should'][]['term']['brand_name'] = 'axe';
$filters['bool']['should'][]['term']['category_name'] = 'axe';
                      //   ^---- Note the new []
```

**Comment by JugalSA** [comment]
Thanks!! 
That worked like a charm...
I want to know that if i have 100 results for a particular search with their individual elasticsearch score.
And i want to take only 60% of the highest score document for every search. How to do that in ES.
"Size" parameter wont work in my case because search is dynamic.
So lets say i have a document with highest score 2.1480842 then i want documents which has score greater than 1.28885052...

**Comment by polyfractal** [comment]
>I want to know that if i have 100 results for a particular search with their individual elasticsearch score.

I'm not sure I understand this question?  The search hits always come back with the score for that document.

> And i want to take only 60% of the highest score document for every search. How to do that in ES.

I'm afraid there is no real way to do this server-side.  Scores are always calculated in the context of the search results.  Meaning it is basically impossible to compare the scores between two queries, because the factors that affect the score change for each query (term and document frequencies, etc).

In terms of technical aspect, there is no way to ask ES to provide 60%, because it only maintains priority queues which are sized to fit the `size` + `from` parameter.  To calculate 60%, Elasticsearch would have to keep all doc scores resident in memory to know where the 60% spot lands.  Which would be very expensive.

Theoretically, the only way to do this is to request *all* search hits by setting an enormous side, then pruning the results client-side.  But that's equally expensive for ES, and forces your app to handle potentially very large responses.

Basically, it's not really possible, nor is it a good idea :)

**Comment by Noctis17** [comment]
Hello sir @polyfractal 

Thank you for your answer, it has given me a guide how to do aggregation in elasticsearch to php
But I got a quick question, how about if I want to do functionalities such as "sum"?

And whenever I perform the code:

$params = [
			'index' => 'mmi.webs',
			'type'	=> '_doc',
			'body'  => [
				'aggs' => [
					'art_ad_val_sum' => [
						'terms' => [
							'field' => 'art_ad_val'
						]
					]
				]
			]
		];

it always give me an error message

"{"error":{"root_cause":[{"type":"parsing_exception","reason":"request does not support [aggs]","line":1,"col":2}],"type":"parsing_exception","reason":"request does not support [aggs]","line":1,"col":2},"status":400}"

Any help for this one please?
